File: sdk/typescript/src/bunqueue/dedup-debounce.ts
/**
 * Bunqueue Simple Mode — deduplication and debounce defaults merger.
 * 1:1 port of src/client/bunqueue/dedupDebounce.ts, including the dedup id
 * derivation algorithm: `${name}:${JSON.stringify(data)}`.
 */

import type { JobOptions } from '../types.js';
import type { BunqueueDebounceConfig, BunqueueDeduplicationConfig } from './types.js';

export class DedupDebounceMerger {
  private readonly dedup: BunqueueDeduplicationConfig | null;
  private readonly debounce: BunqueueDebounceConfig | null;

  constructor(dedup: BunqueueDeduplicationConfig | null, debounce: BunqueueDebounceConfig | null) {
    this.dedup = dedup;
    this.debounce = debounce;
  }

  get active(): boolean {
    return this.dedup !== null || this.debounce !== null;
  }

  merge(name: string, opts?: JobOptions, data?: unknown): JobOptions | undefined {
    if (!this.active) return opts;
    const merged: JobOptions = { ...opts };
    if (this.dedup && !merged.deduplication) {
      const dataKey = data !== undefined ? JSON.stringify(data) : '';
      merged.deduplication = {
        id: `${name}:${dataKey}`,
        ttl: this.dedup.ttl ?? 3600000,
        extend: this.dedup.extend,
        replace: this.dedup.replace,
      };
    }
    if (this.debounce && !merged.debounce) {
      merged.debounce = {
        id: name,
        ttl: this.debounce.ttl,
      };
    }
    return merged;
  }
}

File: sdk/python/bunqueue/simple/dedup_debounce.py
"""Deduplication & debounce defaults merger (mirror of dedupDebounce.ts).

Jobs with the same name + data derive the same deduplication id
(``name:JSON(data)``, compact separators to match JSON.stringify).
"""

from __future__ import annotations

import json
from typing import Any, Dict, Optional


class DedupDebounceMerger:
    def __init__(
        self,
        dedup: Optional[Dict[str, Any]],
        debounce: Optional[Dict[str, Any]],
    ) -> None:
        self._dedup = dedup
        self._debounce = debounce

    @property
    def active(self) -> bool:
        return self._dedup is not None or self._debounce is not None

    def merge(self, name: str, opts: Dict[str, Any], data: Any) -> Dict[str, Any]:
        if not self.active:
            return opts
        merged = dict(opts)
        if self._dedup is not None and "deduplication" not in merged:
            data_key = (
                json.dumps(data, separators=(",", ":")) if data is not None else ""
            )
            merged["deduplication"] = {
                "id": f"{name}:{data_key}",
                "ttl": self._dedup.get("ttl", 3600000),
                "extend": self._dedup.get("extend"),
                "replace": self._dedup.get("replace"),
            }
        if self._debounce is not None and "debounce" not in merged:
            merged["debounce"] = {"id": name, "ttl": self._debounce["ttl"]}
        return merged
